elytron/WFCORE-4058-automatic-keystore-detection.adoc

```diff
@@ -0,0 +1,92 @@
+= Automatic detection of file based KeyStore types.
+:author:            Maritn Mazanek
+:email:             mmazanek@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+* If a user does not specify the type of a file based KeyStore can we auto detect the type?
+
+* There is most likely a finite set of types we would want to automatically recognise, possibly we could inspect the
+contents ourselves to detect different types - alternative attempt initialisation of different KeyStore objects until one succeeds. With the latter we would need to be careful one doesn't silently succeed when it should have failed.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/EAP7-675[EAP7-675]
+* https://issues.jboss.org/browse/ELY-1650[ELY-1650]
+* https://issues.jboss.org/browse/WFCORE-4058[WFCORE-4058]
+
+//=== Related Issues
+
+//* https://issues.jboss.org/browse/WFLY[WFLY-XXXX]
+
+=== Dev Contacts
+
+* mailto:mmazanek@redhat.com[Martin Mazanek]
+
+=== QE Contacts
+
+* TBD
+
+=== Affected Projects or Components
+
+* WildFly, Security
+
+== Requirements
+
+=== Hard Requirements
+
+* Point of this change is to be able to load keystore without specifying its type.
+
+== Implementation Plan
+
+* Finite set of supported keystore types for automatic detection: JKS, JCEKS, PKCS12, BKS, BCFKS, UBER
+* Some types are fairly easily recognizable, as they start with magic number: JKS (0xfeedfeed), JCEKS (0xcececece) - although if a keystore starts with such number it doesn't necessarily mean that it is that type
+* Detect types of encrypted keystores? They may be a good candidate for try until success (let the provider do the decryption for us)
+
+* The idea is to prioritize different techniques with a priority
+** Magic number - check if the store contains one of known magic numbers - if yes, then try to load it and see if it is really that keystore
+** File extension - try to parse as a keystore based on the file extension
+** Try remaining supported types in a specific order / don't try anything and fail
+
+* Recognizable patterns:
+
+** JKS
+*** Starts with magic number 0xfeedfeed
+*** Version number - 0x00000002 or 0x00000001
+
+** JCEKS
+*** Starts with magic number 0xcececece
+*** Version number - 0x00000002 or 0x00000001
+
+** BKS
+*** Starts with version number 0x00000002 - older version (BKS-v1) has version number 0x00000001
+*** Salt length - looking at the BC provider source I found out its 20, hopefully it't the only case - 0x00000014
+*** Salt - 20 bytes
+*** Iteration count - int with value 1024 - 2047 - 0x00000400 - 0x000007ff - top 21 bits are 0
+
+** BCFKS
+*** Whole file encrypted - needs decryption
+
+** PKCS12
+*** ASN.1 encoding - starts with SEQUENCE - tag 0x30
```

Comment:
If this should be in "specific order", you should probably move PKCS12 between JCEKS and BKS.

Reply:
Given PKCS12 is default in JDK9+ and JCEKS is deprecated, I vote for putting it even higher, maybe between JKS and JCEKS :)

Reply:
@nekdozjam To detect PKCS12 do you need to decrypt it first or was it detectable from the raw file?

I would propose those that don't need decryption come first in the list, if anything needs decryption it should then follow.

Reply:
Yeah, this list was just to make notes about each type. And as the JKS, JCEKS and PKCS12 all have some unique patterns at the start of the file so they don't need to be blindly tried out.

Reply:
So maybe the two resolution strategies become: -

1. Defined Type
2. Signature Test -> Extension -> Brute Force

In the second resolution strategy if at any stage the load fails it can fail over to the next in the stage.

Reply:
IMO I don't like brute force unless absolutely necessary - we can still implement signature checks in a way that allows us to fall back to brute force if the signature detection was invalid - the signature check could identify the most likely file type and use that as the highest priority in the brute force loop.

But we should also keep in mind the brute force loop could fail for other reasons such as an invalid password being specified.